src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1241,8 +1242,17 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
                     return true;
 
                 case JsonValueKind.Number:
-                    // JSON numbers are equal if their raw representations are equal.
-                    return left.GetRawValue().Span.SequenceEqual(right.GetRawValue().Span);
+                    ReadOnlySpan<byte> leftRawValue = left.GetRawValue().Span;
+                    ReadOnlySpan<byte> rightRawValue = right.GetRawValue().Span;
+
+                    int bytesConsumed;
+                    if (Utf8Parser.TryParse(leftRawValue, out decimal leftNumber, out bytesConsumed) && bytesConsumed == leftRawValue.Length &&
+                        Utf8Parser.TryParse(rightRawValue, out decimal rightNumber, out bytesConsumed) && bytesConsumed == rightRawValue.Length)
+                    {
+                        return leftNumber == rightNumber;
+                    }
+
+                    return leftRawValue.SequenceEqual(rightRawValue);
```

Comment:
@tannergooding do we have any internal numeric parsers available that support arbitrarily large decimal representations that I could use in this case?

Reply:
The internal `NumberBuffer` type is used by all the built-in numeric types (although it may not be used by BigInteger yet). It basically parses the strings to a common ASCII format that includes some minimal information about scale, etc. Notably, I don't recall off the top of my head how we represent trailing zeros for decimal, since it can represent both `1.0` and `1.00`.

It's probably not much more efficient than simply comparing the two numeric literals, however. I'd expect it'd be faster/simpler to instead compare the two strings getting the index of the first mismatching character, then specially handling trailing zeros if the mismatching character is the decimal point (and leading zeros if relevant, but AFAIR those aren't valid for JSON number literals?)

Reply:
That is, you could likely implement this as:
1. IndexOfAnyExcept('0') - skip leading zeros
2. CommonPrefixLength - from that index, get index of first mismatch
3. Check for `.` and then trailing zeros, again using `IndexOfAnyExcept('0')`

This would be vectorized, handle leading/trailing differences trivially, and allow you to special case other things like `-0` vs `+0` as relevant, without needing to ever produce an actual number.

There is, however, nuance, in that some numbers may compare as equal given a particular format since there are many ways to represent `pi` as `float`, given it has limited precision. But that doesn't seem to be of interest to this method, rather only that the "infinitely precise" number described matches

Reply:
FWIW, leading zeros are disallowed in JSON unless the ipart itself equals zero.  I'm not sure whether we support them in our parsing, though.

Reply:
Yes, I explicitly called that out in the first comment. The general point being that this could be done fairly trivially with zero actual parsing required, just comparing the numeric literals and accounting for the few edges that make values equivalent.

It's a simple, but very extensible approach that keeps things efficient.

Reply:
It would also need to account for possible exponents, but I think the key idea stands.

Reply:
I updated the implementation using a bespoke, arbitrary precision equality comparison. Added more test coverage and fuzzed the implementation using FsCheck for good measure.

Reply:
> and fuzzed the implementation using FsCheck for good measure

Should we add something to the fuzzing suite @MihaZupan set up for the repo?

Reply:
Possibly, can we configure SharpFuzz to accept pairs of input files? 

Using just one input and then checking for reflexivity isn't particularly interesting from the perspective of correctness since it would always be trivially true in this particular case. SharpFuzz seems better suited when validating implementations from a security perspective, in this case I'm interested in the correctness angle where property-based testing is more effective.

Reply:
One possible plan here is to augment the existing [JsonDocument fuzzer](https://github.com/dotnet/runtime/blob/main/src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/JsonDocumentFuzzer.cs) with a deep equals check once [`JsonElement.DeepEquals` has been made public](https://github.com/dotnet/runtime/issues/33388#issuecomment-2181174884).